db/description/collection.go

```diff
@@ -0,0 +1,214 @@
+// Copyright 2023 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package description
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/ipfs/go-datastore/query"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/core"
+	"github.com/sourcenetwork/defradb/datastore"
+)
+
+func SaveCollection(
+	ctx context.Context,
+	txn datastore.Txn,
+	desc client.CollectionDescription,
+) (client.CollectionDescription, error) {
+	buf, err := json.Marshal(desc)
+	if err != nil {
+		return client.CollectionDescription{}, err
+	}
+
+	key := core.NewCollectionKey(desc.ID)
+	err = txn.Systemstore().Put(ctx, key.ToDS(), buf)
+	if err != nil {
+		return client.CollectionDescription{}, err
+	}
+
+	idBuf, err := json.Marshal(desc.ID)
+	if err != nil {
+		return client.CollectionDescription{}, err
+	}
+
+	nameKey := core.NewCollectionNameKey(desc.Name)
+	err = txn.Systemstore().Put(ctx, nameKey.ToDS(), idBuf)
+	if err != nil {
+		return client.CollectionDescription{}, err
+	}
```

Comment:
suggestion: I would do the opposite here. I would store the description on the name key instead of the collection number key because when querying, users will do so using the collection name. This will prevent the extra key lookup on queries which is where speed is most important.

Reply:
I do disagree with you here, for the following reasons:

- Name is mutable, and if the description is stored against it, renaming the collection becomes more expensive and error prone as we have to move the definition and re-index everything.
- The performance cost should be very very small, it is an extra single-point lookup.
- I find it conceptually to make more sense to store the thing against it's fixed ID, and index that, than to store it against a mutable property on it, and then have everything else reference the mutable property.

I think the maintainability of storing against the ID far outweigh the performance gains here.

Reply:
Ah, I should have noted this one in the PR description I think.